File: Classes/Tabs/01/CourseTableViewController.m
//
//  CourseTableViewController.m
//  app
//
//  Created by Jesse Hu on 10/18/14.
//  Copyright (c) 2014 KZ. All rights reserved.
//

#import "CourseTableViewController.h"

@interface CourseTableViewController ()

@end

@implementation CourseTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    // Return the number of rows in the section.
    return [self.courses count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:@"cell"];
    }
    
    NSDictionary *courseDictionary = [self.courses objectAtIndex:[indexPath row]];
    cell.textLabel.text = [courseDictionary objectForKey:@"course_number"];
    cell.detailTextLabel.text = [courseDictionary objectForKey:@"course_title"];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
/*
    course = {
        "course_number": "89S",
        "course_title": "First-Year Seminar",
        "crse_id": "000064",
        "subject_code": "AAAS",
        "subject_desc": "African and African American S"
    }
 */
    
    NSMutableDictionary *course = [[NSMutableDictionary alloc] initWithDictionary:[self.courses objectAtIndex:[indexPath row]] copyItems:YES];
    [course setValue:[self.subjectDictionary objectForKey:@"code"] forKey:@"subject_code"];
    [course setValue:[self.subjectDictionary objectForKey:@"desc"] forKey:@"subject_desc"];
    [self.delegate courseSelected:course];
    [self.navigationController popToRootViewControllerAnimated:YES];
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: SwiftParseChat/CourseTableViewController.swift
//
//  CourseTableViewController.swift
//  SwiftParseChat
//
//  Created by Jesse Hu on 3/9/15.
//  Copyright (c) 2015 Jesse Hu. All rights reserved.
//

import UIKit

protocol CourseTableViewControllerDelegate {
    func didSelectCourse(course: [String: String])
}

class CourseTableViewController: UITableViewController {
    
    var subject: NSDictionary!
    var courses: NSArray!
    var delegate: GroupSelectTableViewControllerDelegate!
    var selectedCourse: [String: String]!
    var subjectVC: SubjectTableViewController!
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }

    // MARK: - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.courses.count
    }

    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath) as UITableViewCell

        if let course = self.courses[indexPath.row] as? [String: String] {
            cell.textLabel?.text = course["course_number"]
            cell.detailTextLabel?.text = course["course_title"]
        }
        
        return cell
    }
    
    // MARK: - Table view delegate
    
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        tableView.deselectRowAtIndexPath(indexPath, animated: true)

        /* Retrieve course and append subject code and name */
        if let course = self.courses[indexPath.row] as? [String: String] {
            self.selectedCourse = course
            if let code = self.subject["code"] as? String {
                self.selectedCourse["subject_code"] = code
            }
            if let desc = self.subject["desc"] as? String {
                self.selectedCourse["subject_desc"] = desc
            }
            
            self.performSegueWithIdentifier("courseToGroupsSegue", sender: self)
        }
    }
    
    // MARK: - Navigation
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "courseToGroupsSegue" {
            let groupSelectVC = segue.destinationViewController as GroupSelectTableViewController
            groupSelectVC.delegate = self.delegate
            groupSelectVC.course = self.selectedCourse
            groupSelectVC.subjectVC = self.subjectVC
        }
    }

}
